File: example/android/awn_core/src/main/java/me/carda/awesome_notifications/core/exceptions/ExceptionFactory.java
package me.carda.awesome_notifications.core.exceptions;

import androidx.annotation.NonNull;

import me.carda.awesome_notifications.core.broadcasters.receivers.AwesomeExceptionReceiver;

public class ExceptionFactory {

    public static String TAG = "ExceptionFactory";

    // ************** SINGLETON PATTERN ***********************

    private static ExceptionFactory instance;

    private ExceptionFactory(){}

    public static ExceptionFactory getInstance() {
        if (instance == null)
            instance = new ExceptionFactory();
        return instance;
    }

    /// **************  FACTORY METHODS  *********************

    public AwesomeNotificationsException createNewAwesomeException(
            @NonNull String className,
            @NonNull String code,
            @NonNull String message,
            @NonNull String detailedCode
    ) {
        return createNewAwesomeException(
                className,
                new AwesomeNotificationsException(
                        code,
                        message,
                        detailedCode));
    }

    public AwesomeNotificationsException createNewAwesomeException(
            @NonNull String className,
            @NonNull String code,
            @NonNull String message,
            @NonNull String detailedCode,
            @NonNull Exception originalException
    ) {
        return createNewAwesomeException(
                className,
                new AwesomeNotificationsException(
                        code,
                        message,
                        detailedCode,
                        originalException));
    }

    public AwesomeNotificationsException createNewAwesomeException(
            @NonNull String className,
            @NonNull String code,
            @NonNull String detailedCode,
            @NonNull Exception e
    ) {
        return createNewAwesomeException(
                className,
                new AwesomeNotificationsException(
                        code,
                        String.format("%s", e.getLocalizedMessage()),
                        detailedCode,
                        e));
    }

    public void registerNewAwesomeException(
            @NonNull String className,
            @NonNull String code,
            @NonNull String message,
            @NonNull String detailedCode
    ) {
        registerAwesomeException(
                className,
                new AwesomeNotificationsException(
                        code,
                        message,
                        detailedCode));
    }

    public void registerNewAwesomeException(
            @NonNull String className,
            @NonNull String code,
            @NonNull String message,
            @NonNull String detailedCode,
            @NonNull Exception originalException
    ) {
        registerAwesomeException(
                className,
                new AwesomeNotificationsException(
                        code,
                        message,
                        detailedCode,
                        originalException));
    }

    public void registerNewAwesomeException(
            @NonNull String className,
            @NonNull String code,
            @NonNull String detailedCode,
            @NonNull Exception originalException
    ) {
        registerAwesomeException(
                className,
                new AwesomeNotificationsException(
                        code,
                        String.format("%s", originalException.getLocalizedMessage()),
                        detailedCode,
                        originalException));
    }

    /// **************  FACTORY METHODS  *********************

    private AwesomeNotificationsException createNewAwesomeException(
            @NonNull String className,
            @NonNull AwesomeNotificationsException exception
    ) {
        AwesomeExceptionReceiver
                .getInstance()
                .notifyNewException(className, exception);
        return exception;
    }

    private void registerAwesomeException(
            @NonNull String className,
            @NonNull AwesomeNotificationsException exception
    ) {
        AwesomeExceptionReceiver
                .getInstance()
                .notifyNewException(className, exception);
    }
}

File: reference/awesome_notifications/ios/AwnCore/Classes/exceptions/ExceptionFactory.swift
//
//  ExceptionFactory.swift
//  awesome_notifications
//
//  Created by CardaDev on 07/04/22.
//

import Foundation


public class ExceptionFactory {
    
    static let TAG:String = "ExceptionFactory"
    
    // ************** SINGLETON PATTERN ***********************
    
    static var instance:ExceptionFactory?
    public static var shared:ExceptionFactory {
        get {
            ExceptionFactory.instance = ExceptionFactory.instance ?? ExceptionFactory()
            return ExceptionFactory.instance!
        }
    }
    private init(){}
    
    // ************** FACTORY METHODS ***********************
    
    public func createNewAwesomeException(
                className:String,
                code:String,
                message:String,
                detailedCode:String
    ) -> AwesomeNotificationsException {
        return createNewAwesomeException(
                fromClassName: className,
                withAwesomeException:
                    AwesomeNotificationsException(
                        className: className,
                        code: code,
                        message: message,
                        detailedCode: detailedCode))
    }

    public func createNewAwesomeException(
            className:String,
            code:String,
            message:String,
            detailedCode:String,
            exception:Error
    ) -> AwesomeNotificationsException {
        return createNewAwesomeException(
                fromClassName: className,
                withAwesomeException:
                    AwesomeNotificationsException(
                        className: className,
                        code: code,
                        message: exception.localizedDescription,
                        detailedCode: detailedCode))
    }

    public func createNewAwesomeException(
            className:String,
            code:String,
            detailedCode:String,
            originalException:Error
    ) -> AwesomeNotificationsException {
        return createNewAwesomeException(
                fromClassName: className,
                withAwesomeException:
                    AwesomeNotificationsException(
                        className: className,
                        code: code,
                        message: originalException.localizedDescription,
                        detailedCode: detailedCode))
    }

    public func registerNewAwesomeException(
            className:String,
            code:String,
            message:String,
            detailedCode:String
    ) {
        _ = createNewAwesomeException(
                fromClassName: className,
                withAwesomeException:
                    AwesomeNotificationsException(
                        className: className,
                        code: code,
                        message: message,
                        detailedCode: detailedCode))
    }

    public func registerNewAwesomeException(
            className:String,
            code:String,
            message:String,
            detailedCode:String,
            originalException:Error
    ) {
        registerAwesomeException(
                fromClassName: className,
                withAwesomeException:
                    AwesomeNotificationsException(
                        className: className,
                        code: code,
                        message: message,
                        detailedCode: detailedCode))
    }

    public func registerNewAwesomeException(
            className:String,
            code:String,
            detailedCode:String,
            originalException:Error
    ) {
        registerAwesomeException(
            fromClassName: className,
            withAwesomeException:
                AwesomeNotificationsException(
                    className: className,
                    code: code,
                    message: originalException.localizedDescription,
                    detailedCode: detailedCode))
    }

    /// **************  FACTORY METHODS  *********************

    private func createNewAwesomeException(
        fromClassName className:String,
        withAwesomeException awesomeException:AwesomeNotificationsException
    ) -> AwesomeNotificationsException {
        AwesomeExceptionReceiver
                .shared
                .notifyExceptionEvent(
                    fromClassName: className,
                    withAwesomeException: awesomeException)
        return awesomeException
    }

    private func registerAwesomeException(
        fromClassName className:String,
        withAwesomeException awesomeException:AwesomeNotificationsException
    ) {
        AwesomeExceptionReceiver
                .shared
                .notifyExceptionEvent(
                    fromClassName: className,
                    withAwesomeException: awesomeException)
    }
}
